### PR TITLE
Initial MultiMC Forge and MC version support

### PR DESCRIPTION
**Pull request by geferon**
I've decided to add MultiMC support, where the packwiz installer will automatically update the MC and Forge versions according to the packs metadata.
Right now this only supports Forge, but I will see if I can add support for Fabric in the future.
I've tested this feature locally and it works perfectly fine, if Forge's version or MC's version differs, the installer will automatically update the JSON file of the pack with the new versions and MultiMC will download and install them accordingly, as Packwiz installer gets run in the pre-launch phase, and it doesn't get overwriten or anything by it, so it runs just fine.
If there's something about the code you don't like or that you'd like me to modify please tell, and I will try to change it to your liking.

### CODE REVIEW

**Comment by comp500** [comment]
Could you add a dialog box asking the user if they want to update? Should have 3 buttons, allowing the user to cancel the launch, continue without changing game/loader version, and automatically update the version.

@EcoBuilder13 you mentioned you were going to implement this, did you get anywhere with that?

**Comment by geferon** [comment]
Sure thing! I'll get that started and update the PR!

**Comment by geferon** [comment]
That should do, I think?
I added a method to IUserInterface called showCustomDialog
Basically taking three parameters which would be the usual for a JOptionPane.showOptionDialog method.
I dunno what your convention on the regard is, or if you'd prefer using a method like... "showUpdateConfirmationDialog" or such instead with an Enum as a return?

**Comment by geferon** [comment]
Oh I also added support for Fabric while I was at it

**Comment by comp500** [comment]
> "showUpdateConfirmationDialog" or such instead with an Enum as a return?
I'd prefer this, particularly since it allows the CLI to implement it in future (or use flags/etc)

**Comment by geferon** [comment]
Alrighty, I'll refactor it when I have a chance at it tomorrow

**Comment by geferon** [comment]
Okay that should do it

**Comment by geferon** [comment]
Doesn't MultiMC automatically handle mappings on launch? As far as I've seen after testing with this, MultiMC does re-dl the mappings on launch and update the json file accordingly.
Now, for LWJGL I have no idea.

File: src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt [hunk 1]
```diff
@@ -59,7 +61,7 @@ class UpdateManager internal constructor(private val opts: Options, val ui: IUse
 		checkOptions()
 
 		ui.submitProgress(InstallProgress("Loading manifest file..."))
-		val gson = GsonBuilder().registerTypeAdapter(Hash::class.java, Hash.TypeHandler()).create()
+		val gson = GsonBuilder().registerTypeAdapter(Hash::class.java, Hash.TypeHandler()).setPrettyPrinting().create()
```
Comment (comp500) [inline src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt:64]: This increases the filesize quite a bit for large packs, not sure if it's necessary though it does make it easier to debug :P

File: src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt [hunk 2]
```diff
@@ -97,6 +99,105 @@ class UpdateManager internal constructor(private val opts: Options, val ui: IUse
 			handleCancellation()
 		}
 
+		// MultiMC MC and loader version checker
```
Comment (comp500) [inline src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt:102]: Could you move this to a separate class/file? Should make it easier to merge with some of the other pending changes I've got

File: src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt [hunk 3]
```diff
@@ -97,6 +99,105 @@ class UpdateManager internal constructor(private val opts: Options, val ui: IUse
...
+			Log.info("Loaded MultiMC config")
+
+			// We only support format 1, if it gets updated in the future we'll have to handle that
+			// There's only version 1 for now tho, so that's good
+			if (multimcManifest["formatVersion"].asInt != 1) {
+				ui.showErrorAndExit("Invalid MultiMC format version")
+			}
+
+			var manifestModified = false
+			val modLoaders = hashMapOf("net.minecraftforge" to "forge", "net.fabricmc.fabric-loader" to "fabric")
```
Comment (comp500) [inline src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt:126]: Could also support quilt, liteloader?

File: src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt [hunk 4]
```diff
@@ -97,6 +99,105 @@ class UpdateManager internal constructor(private val opts: Options, val ui: IUse
...
+			var modLoaderFound = false
+			var modLoader: String? = null
+			var modLoaderOldVer: String? = null
+			var mcOldVer = "Unknown"
+			val components = multimcManifest["components"].asJsonArray
+			for (componentObj in components) {
+				val component = componentObj.asJsonObject
+
+				val version = component["version"].asString
+				when (component["uid"].asString) {
```
Comment (comp500) [inline src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt:137]: Don't need a `when` for this, might as well turn it into an `if`

Comment (comp500) [reply src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt:137]: Actually it doesn't even need to be special-cased, you might as well just make it part of `modLoaders` and allow updating multiple

File: src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt [hunk 5]
```diff
@@ -97,6 +99,105 @@ class UpdateManager internal constructor(private val opts: Options, val ui: IUse
...
+			// If we can't find the mod loader in the MultiMC file, we add it
+			if (!modLoaderFound) {
+				for ((loaderClass, loader) in modLoaders) {
+					if (pf.versions?.containsKey(loader) ?: false) {
+						modLoader = loader
+						break
+					}
+				}
+				// If we can't find it in the modpack pack file, something is wrong, we'll stop here
+				if (modLoader == null) ui.showErrorAndExit("Mod Loader not found in modpack pack file")
```
Comment (comp500) [inline src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt:167]: Entirely possible to have a packwiz "modpack" with no loader; maybe someone has a pack to distribute their resource packs or the loader is unsupported (which is fine, just continue anyway)

File: src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt [hunk 6]
```diff
@@ -97,6 +99,105 @@ class UpdateManager internal constructor(private val opts: Options, val ui: IUse
...
+				multimcManifestFile.writeText(gson.toJson(multimcManifest))
+				Log.info("Updated modpack Minecraft and/or ${modLoader?.replaceFirstChar { it.uppercase() }} version")
+			}
+
+			if (ui.cancelButtonPressed) {
+				showCancellationDialog()
+				handleCancellation()
+			}
+		} else {
+			Log.warn("MultiMC installation not detected... Tried looking for it in $multimcManifestPath")
```
Comment (comp500) [inline src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt:198]: Shouldn't be a warning, packwiz-installer does not need to be run under MultiMC (can be run standalone, I'm working on vanilla launcher support)

File: src/main/kotlin/link/infra/packwiz/installer/ui/gui/GUIHandler.kt [hunk 7]
```diff
@@ -166,6 +166,59 @@ class GUIHandler : IUserInterface {
...
+				message.append("</font></li>")
+			}
+			message.append("</ul><br>" +
+					"Would you like to update the versions, launch without updating, or cancel the launch?")
+
+
+			val options = arrayOf("Cancel", "Continue anyways", "Update")
+			val result = JOptionPane.showOptionDialog(frmPackwizlauncher,
+					message,
+					"MultiMC versions updated",
```
Comment (comp500) [inline src/main/kotlin/link/infra/packwiz/installer/ui/gui/GUIHandler.kt:208]: Should be past tense (the file hasn't been updated yet)

File: src/main/kotlin/link/infra/packwiz/installer/ui/gui/GUIHandler.kt [hunk 8]
```diff
@@ -166,6 +166,59 @@ class GUIHandler : IUserInterface {
...
 
+	override fun showUpdateConfirmationDialog(oldVersions: List<Pair<String, String?>>, newVersions: List<Pair<String, String?>>): IUserInterface.UpdateConfirmationResult {
+		assert(newVersions.isNotEmpty())
+		val future = CompletableFuture<IUserInterface.UpdateConfirmationResult>()
+		EventQueue.invokeLater {
+			val oldVersIndex = oldVersions.map { it.first to it.second }.toMap()
+			val newVersIndex = newVersions.map { it.first to it.second }.toMap()
+			val message = StringBuilder()
+			message.append("<html>" +
+					"The modpacks version has been updated.<br>" +
```
Comment (comp500) [inline src/main/kotlin/link/infra/packwiz/installer/ui/gui/GUIHandler.kt:177]: Maybe state something like "This modpack update uses newer versions of the following:"

**Review by comp500 — CHANGES_REQUESTED** [review]

**Comment by comp500** [comment]
> lack of Quilt support and lack of loader switching support (ex: converting a Fabric modpack to Quilt).

Agreed, this would definitely be appreciated in the PR.

> LWGJL is a little bit harder to take care of and has more edge cases. It gets automatically updated when updating from 1.18.x to 1.19.x in the UI but otherwise I'm not sure how it handles it.

I did some testing; it looks like MultiMC will automatically handle this for you, even if you change the game version by editing `mmc-pack.json` it will get changed to the correct version.

> To my knowledge Intermediary Mappings updates whenever you update the game version and has the same value of the game version.

If you change it from `mmc-pack.json` it looks like it complains about the versions not matching, rather than automatically changing it. If you remove the `net.fabricmc.intermediary` component though, it will retrieve the correct version when using Quilt/Fabric - so I think the fix would be to just remove the `net.fabricmc.intermediary` component whenever changing the Minecraft version.

**Comment by geferon** [comment]
Just pushed changes for most of the review, I just need your input on two of them.

**Comment by geferon** [comment]
On the review there are two pending comments yeah.

**Comment by comp500** [comment]
You might need to complete the review for me to see them?

**Comment by geferon** [comment]
You should be able to open the review and see it, it only allows me to cancel the review, you're the only one allowed to approve the review or request more changes.

**Comment by geferon** [comment]
I believe you were right but at the same time if you click on those two unresolved conversations you should be able to see my comment on them and being able to comment on it.
Anyways, I submitted it for review.

File: src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt (hunk 1 above)
Comment (comp500) [reply src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt:64]: Yeah, please do :)

File: src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt (hunk 2 above)
Comment (comp500) [reply src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt:102]: Just stick it in the `installer` package for now, I can refactor it later. Returning a bool or enum would work, ideally only access the cancellation code from `UpdateManager` itself. `LauncherUtils` would be good, particularly if I want to add support for other launchers in the future.

**Comment by comp500** [comment]
I see them now - they weren't there before! The UI is a bit strange, I think if you add comments from the changes tab it will require "submitting" the review but it's not super clear when you need to do that.

File: src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt (hunk 2 above)
Comment (geferon) [reply src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt:102]: It's what imma do for now, if you want me to change it, comment on this and if not you can just mark it as resolved and merge the PR.

**Comment by geferon** [comment]
Oh huh, that's unintuitive, I pushed all changes

File: src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt (hunk 2 above)
Comment (geferon) [reply src/main/kotlin/link/infra/packwiz/installer/UpdateManager.kt]: It's what imma do for now, if you want me to change it, comment on this and if not you can just mark it as resolved and merge the PR.

**Pull request merged**